l2-contracts/contracts/ConsensusRegistry.sol

```diff
@@ -0,0 +1,444 @@
+
+        _ensureValidatorSnapshot(node);
+        node.validatorLatest.pubKey = _pubKey;
+        node.validatorLatest.pop = _pop;
+
+        emit NodeValidatorKeyChanged(_nodeOwner, _pubKey, _pop);
+    }
+
+    /// @notice Changes the attester's public key of a node in the registry.
+    /// @dev Only callable by the contract owner or the node owner.
+    /// @dev Verifies that the node owner exists in the registry.
+    /// @param _nodeOwner The address of the node's owner whose attester public key will be changed.
+    /// @param _pubKey The new ECDSA public key to assign to the node's attester.
+    function changeAttesterKey(
+        address _nodeOwner,
+        Secp256k1PublicKey calldata _pubKey
+    ) external onlyOwnerOrNodeOwner(_nodeOwner) {
+        _verifyInputSecp256k1PublicKey(_pubKey);
+        _verifyNodeOwnerExists(_nodeOwner);
+        Node storage node = nodes[_nodeOwner];
+        if (_finalizeNodeRemoval(_nodeOwner, node)) {
+            return;
+        }
+
+        bytes32 prevHash = _hashAttesterPubKey(node.attesterLatest.pubKey);
+        delete attesterPubKeyHashes[prevHash];
+        bytes32 newHash = _hashAttesterPubKey(_pubKey);
+        attesterPubKeyHashes[newHash] = true;
+
+        _ensureAttesterSnapshot(node);
+        node.attesterLatest.pubKey = _pubKey;
+
+        emit NodeAttesterKeyChanged(_nodeOwner, _pubKey);
+    }
+
+    /// @notice Adds a new commit to the validator committee.
+    /// @dev Implicitly updates the validator committee by affecting off-chain readers based on the current state of a node's validator attributes:
+    /// - If "validatorsCommit" > "node.validatorLastUpdateCommit", read "node.validatorLatest".
+    /// - If "validatorsCommit" == "node.validatorLastUpdateCommit", read "node.validatorSnapshot".
+    /// @dev Only callable by the contract owner.
+    function commitValidators() external onlyOwner {
+        validatorsCommit++;
+
+        emit ValidatorsCommitted();
+    }
+
+    /// @notice Adds a new commit to the attester committee.
+    /// @dev Implicitly updates the attester committee by affecting off-chain readers based on the current state of a node's attester attributes:
+    /// - If "attestersCommit" > "node.attesterLastUpdateCommit", read "node.attesterLatest".
+    /// - If "attestersCommit" == "node.attesterLastUpdateCommit", read "node.attesterSnapshot".
+    /// @dev Only callable by the contract owner.
+    function commitAttesters() external onlyOwner {
+        attestersCommit++;
+
+        emit AttestersCommitted();
+    }
+
+    function numNodes() public view returns (uint256) {
+        return nodeOwners.length;
+    }
+
+    function _finalizeNodeRemoval(address _nodeOwner, Node storage _node) private returns (bool) {
+        if (
+            _node.attesterSnapshot.pendingRemoval &&
+            _node.validatorSnapshot.pendingRemoval
+        ) {
+            nodeOwners[_nodeOwnerIdx(_nodeOwner)] = nodeOwners[nodeOwners.length - 1];
```

Comment:
Changes here: https://github.com/matter-labs/era-contracts/pull/555/commits/feb76be97a8f8043a6fba30112d38d02fef37107

It's a bit ugly because for updating the index in the moved element I added another mapping to keep track between the index and the node owner address (the latter is needed for accessing the node instance needed for the update).